akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala

```diff
@@ -1368,6 +1368,32 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
   def mapError(pf: PartialFunction[Throwable, Throwable]): javadsl.Flow[In, Out, Mat] =
     new Flow(delegate.mapError(pf))
 
+  /**
+   * Java API
+   *
+   * While similar to [[recover]] this operator can be used to transform an error signal to a different one *without* logging
+   * it as an error in the process. So in that sense it is NOT exactly equivalent to `recover(t => throw t2)` since recover
+   * would log the `t2` error.
+   *
+   * Since the underlying failure signal onError arrives out-of-band, it might jump over existing elements.
+   * This operator can recover the failure signal, but not the skipped elements, which will be dropped.
+   *
+   * Similarly to [[recover]] throwing an exception inside `mapError` _will_ be logged.
+   *
+   * '''Emits when''' element is available from the upstream or upstream is failed and pf returns an element
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes or upstream failed with exception pf can handle
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   */
+  def mapError[E <: Throwable](clazz: Class[E], f: function.Function[E, Throwable]): javadsl.Flow[In, Out, Mat] =
+    mapError {
+      case err if clazz.isInstance(err) ⇒ f(clazz.cast(err))
+    }
```

Comment:
Wouldn't `isAssignableFrom` be better, so you don't have to match on exact exception class but can catch a whole subtree if you like? `mapError(RuntimeException.class, ex -> new SomeOther(ex))`

Reply:
i actually considered that too, but then saw that `isInstance` is used in the implementation of [`recoverWith`]. so i stuck with that for consistency. but i'm certainly amenable to using `isAssignableFrom` instead.

[`recoverWith`]: https://github.com/404-/akka/blob/0166d84190c336d8fb88a3180533df2a87dff162/akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala#L1442

Reply:
a more general observation (for a potential follow-up task), would it make more sense to have a utility method for building Scala `PartialFunction`s from an instance of `Class` and a mapping `Function`? e.g.:

```
static <I, O, A extends I, B extends O> PartialFunction<I, O> pf(final Class<A> c, final Function<A, B> f) {
  return akka.japi.pf.Match.<I, O, A>match(c, f::apply).build();
}

//later (in Java):
  someFlow.mapError(pf(NoSuchElementException.class, (NoSuchElementException e) -> new IllegalArgumentException("boom")))
```

having a utility like this would cut down on the number of "helper" APIs needed for the Java flavor of Akka.

unfortunately, the existing utilities such as  `akka.japi.pf.Match` or `akka.japi.pf.PFBuilder` aren't quite sufficient (and too cumbersome) by themselves due to how the methods are parameterized.

Reply:
We are rather grativating towards specific APIs for Java than such interop-utils, even if it gives more API surface, because it is so much easier to discover for new devs.

The reason I proposed `isAssignableFrom` is that what I think you'd likely do in the Scala api is something like this: 
```
.mapError {
  case ex: RuntimeException => new WhateverExc()
}
```
Which would catch subtypes as well. 

In addition to this it is also how "normal" try-catch-code catches exceptions.

I'm somewhat surprised that's not already how `recoverWith` works in the javadsl, I think we should change that as well.

Reply:
After looking into `isInstance ` a bit more, it does exactly what's desired, per JavaDoc:
> Determines if the specified Object is assignment-compatible with the object represented by this Class. This method is the dynamic equivalent of the Java language instanceof operator.

I added an [extra unit test] which verifies matching on the parent exception type.

[extra unit test]: https://github.com/akka/akka/pull/26310/commits/6138e92339dbd39035b1284a77e644184c304fd1#diff-3b98156e098f2c6b236ff08f1b9d0598R1087

Reply:
Ah, mistake on my part. Great! Thanks for double checking and covering with a test!